src/bookshelf/src/bksw_sec.erl

```diff
@@ -21,123 +22,405 @@
+
+        case Host = wrq:get_req_header("Host", Req0) of
+            undefined -> throw({RequestId, Req0, Context});
+            _         -> ok
+        end,
+
+        % CODE REVIEW: Used in obtaining CredentialScopeDate, which is used for Date, which is used in both verification types
+
+        [AWSAccessKeyId, CredentialScopeDate, Region | _] =
+            case parse_x_amz_credential(Credential) of
+                {error,      _} -> throw({RequestId, Req0, Context});
+                {ok, ParseCred} -> ParseCred
+            end,
+
+        % CODE REVIEW: Date is used in both verification types.
+
+        % https://docs.aws.amazon.com/general/latest/gr/sigv4-date-handling.html
+        DateIfUndefined = wrq:get_req_header("date", Req0),
+        case {_,  Date} = get_check_date(XAmzDate, DateIfUndefined, CredentialScopeDate) of
+            {error,  _} -> throw({RequestId, Req0, Context});
+            {ok,     _} -> ok
+        end,
+
+        % CODE REVIEW: Used in generating Config which is used in both verification types
+        AccessKey = bksw_conf:access_key_id(    Context),
+        SecretKey = bksw_conf:secret_access_key(Context),
+
+        % CODE REVIEW: used in both verification types
+        Headers = process_headers(Headers0),
+
+        % CODE REVIEW: used in both verification types
+        SignedHeaderKeys = parse_x_amz_signed_headers(SignedHeaderKeysString),
+        SignedHeaders = get_signed_headers(SignedHeaderKeys, Headers, []),
+
+        % CODE REVIEW: used in both verification types
+        RawMethod = wrq:method(Req0),
+        Method = list_to_atom(string:to_lower(erlang:atom_to_list(RawMethod))),
+
+        % CODE REVIEW: used in both verification types
+        Path = wrq:path(Req0),
+
+        % CODE REVIEW: used in both verification types
+        Config = mini_s3:new(AccessKey, SecretKey, Host),
+
+        % CODE REVIEW: AltSignedHeaders used in both verification types
+
+        % replace host header with alternate host header
+        AltHost = mini_s3:get_host_toggleport(Host, Config),
+        AltSignedHeaders = [case {K, V} of {"host", _} -> {"host", AltHost}; _ -> {K, V} end || {K, V} <- SignedHeaders],
+
+        % CODE REVIEW: used in both verification types
+        XAmzExpires = list_to_integer(XAmzExpiresString),
+        case XAmzExpires > 1 andalso XAmzExpires < 604800 of
+            true -> ok;
+            _    -> throw({RequestId, Req0, Context})
+        end,
+
+        CalculatedSig = case VerificationType of
+            presigned_url ->
+
+                case wrq:get_qs_value("X-Amz-Algorithm", Req0) of
+                    "AWS4-HMAC-SHA256" -> ok;
+                    _                  -> throw({RequestId, Req0, Context})
+                end,
+
+                case check_signed_headers_common(SignedHeaders, Headers) of
+                    true -> ok;
+                    _    -> throw({RequestId, Req0, Context})
+                end,
```

Comment:
We could check if auth is using presigned method or by using the authorization headers in the malformed_requests function (described above) and include these checks in that function.

Reply:
Done.